### PR TITLE
Add sentinel and sanitize package

### PR DESCRIPTION
**Pull request by louisruch**
### What does this PR do

Adds a new db sentinel and sanitize package to allow go code to validate if a value is a sentinel and to sanitize input from an external source.

### CODE REVIEW

**Comment by hb9cwp** [comment]
> ### What does this PR do
> Adds a new db sentinel and sanitize package to allow go code to validate if a value is a sentinel and to sanitize input from an external source.

So this is not an integration with your other project/product "Sentinel", but another "sentinel" local to Boundary and it's PostgreSQL DB itself. Just an observation from the outside, not wanting to interfere with great development work in any way.

**Comment by louisruch** [comment]
@talanknight PTAL

**Comment by louisruch** [comment]
> > ### What does this PR do
> > Adds a new db sentinel and sanitize package to allow go code to validate if a value is a sentinel and to sanitize input from an external source.
> 
> So this is not an integration with your other project/product "Sentinel", but another "sentinel" local to Boundary and it's PostgreSQL DB itself. Just an observation from the outside, not wanting to interfere with great development work in any way.

Yes your summary is correct, this is not a integration with Hashicorp Sentinel :) sentinel here is https://en.wikipedia.org/wiki/Sentinel_value

File: internal/db/sanitize/doc.go [hunk 1]
```diff
@@ -0,0 +1,3 @@
+// Package sanitize contains a set of functions that sanitizes input received from external
```
Comment (mgaffney) [inline internal/db/sanitize/doc.go:1]: Nice!

File: internal/db/sanitize/sanitize.go [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+// String sanitizes s by replacing all invalid unicode characters as well as the sentinel
+// start character U+FFFE and sentinel end character U+FFFF with the Unicode
+// replacement character U+FFFD.
+//
+// According to the Unicode standard: "If a noncharacter is received in open interchange,
+// an application is not required to interpret it in any way. It is good practice, however,
+// to recognize it as a noncharacter and to take appropriate action, such as replacing it
+// with U+FFFD replacement character."
+// See https://www.unicode.org/versions/Unicode13.0.0/ch23.pdf#G12612.
```
Comment (mgaffney) [inline internal/db/sanitize/sanitize.go:17]: This is great. Thanks for adding it!

File: internal/db/schema/migrations/postgres/10/07_sentinels.up.sql [hunk 3]
```diff
@@ -0,0 +1,36 @@
+begin;
+
+update credential_vault_credential
+   set external_id = concat(external_id, u&'\ffff')
+ where wt_is_sentinel(external_id)
+   and not starts_with(reverse(external_id), u&'\ffff');
```
Comment (mgaffney) [inline internal/db/schema/migrations/postgres/10/07_sentinels.up.sql:6]: Nice use of the `reverse` function :)

File: internal/db/schema/postgres_migration.gen.go [hunk 4]
```diff
@@ -6041,6 +6041,38 @@ create table session_credential_dynamic (
   $$ language plpgsql;
   create trigger revoke_credentials after insert on session_state
     for each row execute procedure revoke_credentials();
+`),
```
Comment (mgaffney) [inline internal/db/schema/postgres_migration.gen.go:6044]: It looks like you need to run `make migrations` again.

File: internal/db/sentinel/doc.go [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+// Boundary defined sentinels and values provided by external systems.
+//
+// All sentinel values are prefixed with the sentinel start character U+FFFE and suffixed
+// with the sentinel end character U+FFFF. Any string that starts with U+FFFE and ends with
+// U+FFFF is a valid sentinel and reserved for use within Boundary.
+//
+// U+FFFE and U+FFFF are special non-characters reserved for internal use in the Unicode
+// standard.
+// See https://www.unicode.org/versions/Unicode13.0.0/ch23.pdf#G12612.
+package sentinel
```
Comment (mgaffney) [inline internal/db/sentinel/doc.go:11]: One of the best package docs I've seen so far.

**Review by mgaffney — APPROVED** [review]
Please run `make migrations` again. Otherwise, this looks great!

File: internal/db/schema/postgres_migration.gen.go (hunk 4 above)
Comment (louisruch) [reply internal/db/schema/postgres_migration.gen.go:6044]: good catch

File: internal/db/sentinel/doc.go (hunk 5 above)
Comment (louisruch) [reply internal/db/sentinel/doc.go:11]: ty :)

**Review by mgaffney — APPROVED** [review]

**Pull request merged**